metagpt/actions/write_code.py

```diff
@@ -135,20 +157,64 @@ async def run(self, *args, **kwargs) -> CodingContext:
         return coding_context
 
     @staticmethod
-    async def get_codes(task_doc, exclude) -> str:
+    async def get_codes(task_doc, exclude, mode="normal") -> str:
+        """
+        Get code snippets based on different modes.
+
+        Attributes:
+            task_doc (Document): Document object of the task file.
+            exclude (str): Specifies the filename to be excluded from the code snippets.
+            mode (str): Specifies the mode, either "normal" or "guide" (default is "normal").
+
+        Returns:
+            str: Code snippets.
+
+        Description:
+        If mode is set to "normal", it returns code snippets for the regular coding phase,
+        i.e., all the code generated before writing the current file.
+
+        If mode is set to "guide", it returns code snippets for incremental development,
+        building upon the existing code in the "normal" mode and adding code for the current file's older versions.
+        """
         if not task_doc:
             return ""
         if not task_doc.content:
             task_doc.content = FileRepository.get_file(filename=task_doc.filename, relative_path=TASK_FILE_REPO)
         m = json.loads(task_doc.content)
-        code_filenames = m.get("Task list", [])
+        code_filenames = m.get(TASK_LIST.key, []) if mode == "normal" else m.get(REFINED_TASK_LIST.key, [])
         codes = []
         src_file_repo = CONFIG.git_repo.new_file_repository(relative_path=CONFIG.src_workspace)
-        for filename in code_filenames:
-            if filename == exclude:
-                continue
-            doc = await src_file_repo.get(filename=filename)
-            if not doc:
-                continue
-            codes.append(f"----- {filename}\n" + doc.content)
+
+        if mode == "guide":
+            src_files = src_file_repo.all_files
+            old_file_repo = CONFIG.git_repo.new_file_repository(relative_path=CONFIG.old_workspace)
+            old_files = old_file_repo.all_files
+            union_files_list = list(set(src_files) | set(old_files))
+            for filename in union_files_list:
+                if filename == exclude:
+                    # Exclude unnecessary code to maintain a clean and focused main.py file, ensuring only relevant and
+                    # essential functionality is included for the project’s requirements
+                    if filename in old_files and filename != "main.py":
+                        # Use legacy code
+                        doc = await old_file_repo.get(filename=filename)
+                    else:
+                        continue
+                    codes.insert(0, f"-----Now, {filename} to be rewritten\n```{doc.content}```\n=====")
+
+                else:
+                    # Use new code
```

Comment:
More comments are needed here, and the design logic is not discernible from the code alone